--- a/src/hotspot/share/jfr/jni/jfrJavaSupport.cpp
+++ b/src/hotspot/share/jfr/jni/jfrJavaSupport.cpp
@@ -435,8 +435,13 @@
   assert(result != nullptr, "invariant");
   DEBUG_ONLY(JfrJavaSupport::check_java_thread_in_vm(THREAD));
   result->set_oop(nullptr); // Initialize result in case klass initialize throws.
+  HandleMark hm(THREAD);
+  Handle receiver(THREAD, args->has_receiver() ? args->receiver() : nullptr);
   InstanceKlass* const klass = static_cast<InstanceKlass*>(args->klass());
   klass->initialize(CHECK);
+  if (receiver.not_null()) {
+    args->set_receiver(receiver);
+  }
   read_field(args, result, static_cast<Thread*>(THREAD));
 }
 
@@ -444,16 +449,20 @@
   assert(args != nullptr, "invariant");
   DEBUG_ONLY(JfrJavaSupport::check_java_thread_in_vm(THREAD));
 
+  HandleMark hm(THREAD);
+  Handle receiver(THREAD, args->has_receiver() ? args->receiver() : nullptr);
   InstanceKlass* const klass = static_cast<InstanceKlass*>(args->klass());
   klass->initialize(CHECK);
+  if (receiver.not_null()) {
+    args->set_receiver(receiver);
+  }
 
   const bool static_field = !args->has_receiver();
   fieldDescriptor fd;
   lookup_field(args, klass, &fd, static_field);
   assert(fd.offset() > 0, "invariant");
 
-  HandleMark hm(THREAD);
-  Handle h_oop(static_field ? Handle(THREAD, klass->java_mirror()) : Handle(THREAD, args->receiver()));
+  Handle h_oop(static_field ? Handle(THREAD, klass->java_mirror()) : receiver);
   write_specialized_field(args, h_oop, &fd, static_field);
 }
 
--- a/src/hotspot/share/jfr/dcmd/jfrDcmds.cpp
+++ b/src/hotspot/share/jfr/dcmd/jfrDcmds.cpp
@@ -291,7 +291,7 @@
   return str;
 }
 
-static const char* read_string_field(oop argument, const char* field_name, TRAPS) {
+static const char* read_string_field(Handle argument, const char* field_name, TRAPS) {
   JavaValue result(T_OBJECT);
   JfrJavaArguments args(&result);
   args.set_klass(argument->klass());
@@ -303,7 +303,7 @@
   return string_oop != nullptr ? get_as_dcmd_arena_string(string_oop) : nullptr;
 }
 
-static bool read_boolean_field(oop argument, const char* field_name, TRAPS) {
+static bool read_boolean_field(Handle argument, const char* field_name, TRAPS) {
   JavaValue result(T_BOOLEAN);
   JfrJavaArguments args(&result);
   args.set_klass(argument->klass());
@@ -314,7 +314,7 @@
   return (result.get_jint() & 1) == 1;
 }
 
-static DCmdArgumentInfo* create_info(oop argument, TRAPS) {
+static DCmdArgumentInfo* create_info(Handle argument, TRAPS) {
   return new DCmdArgumentInfo(
     read_string_field(argument, "name", THREAD),
     read_string_field(argument, "description", THREAD),
@@ -344,14 +344,14 @@
     assert(array->length() == _num_arguments, "invariant");
     return array;
   }
-  objArrayOop arguments = objArrayOop(result.get_oop());
-  assert(arguments != nullptr, "invariant");
+  objArrayHandle arguments(thread, objArrayOop(result.get_oop()));
+  assert(arguments.not_null(), "invariant");
   assert(arguments->is_array(), "must be array");
   const int num_arguments = arguments->length();
   assert(num_arguments == _num_arguments, "invariant");
   prepare_dcmd_string_arena(thread);
   for (int i = 0; i < num_arguments; ++i) {
-    DCmdArgumentInfo* const dai = create_info(arguments->obj_at(i), thread);
+    DCmdArgumentInfo* const dai = create_info(Handle(thread, arguments->obj_at(i)), thread);
     assert(dai != nullptr, "invariant");
     array->append(dai);
   }
